Render self links

terahn opened the pull request:
Implements the feature outlined in #391 

So this code will render a self link if the source and target nodes have the same x and y values. This means that this would also apply to 2 separate nodes if they had the same x and y values. I could pass in the source and target id's to make sure that we are only rendering these links for actual self links, however I think leaving it as is could also be beneficial. 

If, for example, two nodes are so close together that you cannot see the link between them, then maybe rendering one of these "self links" would be useful to help visualize that these two nodes are connected.

![output](https://user-images.githubusercontent.com/23760949/99581536-cc563600-2995-11eb-968c-a59fd94a627e.gif)

Let me know your thoughts and I can make further changes as you see fit.

antoninklopp commented:
Hi @terahn , 

Awesome that you are tackling this issue. 

It provides a good solution to a *self node link*, however, I am not sure that this is what is expected for a graph with *two nodes at the same position*. I think that this rounded self-link really means that this is a *same node link* and not a *same position link*.  
It creates a strange behaviour when you drag a node closer to another and suddenly a link appears when you are at the exact position and disappears right after. 

What are your thoughts on this?   


As a side comment, do you think that we should choose the side where this node link is rendered? 
For example here :
![self-node](https://user-images.githubusercontent.com/26838971/99880398-60810100-2c13-11eb-9a5e-0e64f0083b66.PNG)
There is a lot of place on the left side and the link is still rendered on the right. Do you think that this is something that makes sense and that could be handled?

terahn commented:
> Hi @terahn ,
> 
> Awesome that you are tackling this issue.
> 
> It provides a good solution to a _self node link_, however, I am not sure that this is what is expected for a graph with _two nodes at the same position_. I think that this rounded self-link really means that this is a _same node link_ and not a _same position link_.
> It creates a strange behaviour when you drag a node closer to another and suddenly a link appears when you are at the exact position and disappears right after.
> 
> What are your thoughts on this?

Hey @antoninklopp,

I agree with what you're saying. I think rendering a curved link when two different nodes are close together could be a useful feature to investigate, but I guess it's out of the scope of this particular issue. I'll add some additional logic to make sure that these links are only rendered for actual self-links.

> As a side comment, do you think that we should choose the side where this node link is rendered?

This was something else I was wondering about. I'll investigate!

terahn commented:
- I've updated the logic to ensure that self links are rendered only when the source and target id's match. 
- I have also implemented configuration that allows the user to choose which quadrant the self link is rendered in.

![Screen Shot 2020-12-03 at 1 52 29 PM](https://user-images.githubusercontent.com/23760949/101093145-33b3de80-356f-11eb-812d-30ade71969cf.png)

My understanding of SVG paths is limited so there may be a more elegant solution for controlling the direction in which the self link is rendered, so let me know if there are any improvements I could make.

danielcaldas reviewed (approved):
This looks great! What do you think @antoninklopp?

antoninklopp reviewed (approved):
This looks awesome! Great work!
